### PR TITLE
Nonce and max_age handling with new CookieStore class

### PR DESCRIPTION
**Pull request by joshcanhelp**
### Changes

**Potentially breaking changes:**

- Add enforced nonce checking for ID tokens. Nonce value is stored automatically in cookies (default cookie name `auth0_nonce`) and required to be checked in ID tokens. All checking is handled in the SDK but some managed hosts require whitelisted cookies and/or specific cookie names so this could be breaking.

**Other changes:**

- Add `auth_store` config option for ID token `nonce` handling; defaults to new `CookieStore` class
- Add `max_age` config option for `Auth0` class to set a `max_age` URL parameter on all authorization requests.
- Add `leeway` config option for `Auth0` class to set an ID token time check leeway

### Testing

Also went through a number of manual tests to ensure the default cookie setting and deleting was working. 

- [x] This change adds test coverage
- [x] This change has been developed and tested on PHP 7.1 and 7.2

### Checklist

- [x] All existing and new tests complete without errors.
- [x] The correct base branch is being used.

### CODE REVIEW

File: tests/Auth0Test.php [hunk 1]
```diff
@@ -44,6 +37,24 @@ class Auth0Test extends TestCase
...
+    public function setUp()
+    {
+        parent::setUp();
+        self::$baseConfig = [
+            'domain'        => '__test_domain__',
+            'client_id'     => '__test_client_id__',
+            'client_secret' => '__test_client_secret__',
+            'redirect_uri'  => '__test_redirect_uri__',
+            'store' => false,
+            'auth_store' => new SessionStore(),
```
Comment (joshcanhelp) [inline tests/Auth0Test.php]: Need to use session store to check that values were set

File: tests/Auth0Test.php [hunk 2]
```diff
@@ -27,15 +29,7 @@ class Auth0Test extends TestCase
      *
      * @var array
      */
-    public static $baseConfig = [
```
Comment (joshcanhelp) [inline tests/Auth0Test.php]: Need to move this to `setUp` to use a class for `auth_store`

File: src/Auth0.php [hunk 3]
```diff
@@ -196,6 +198,20 @@ class Auth0
      */
     protected $stateHandler;
 
+    /**
+     * Maximum time allowed between authentication and ID token verification.
+     *
+     * @var integer
+     */
+    protected $maxAge;
```
Comment (joshcanhelp) [inline src/Auth0.php]: Allows a max_age time to be set at the initialization level for all requests.

File: src/Auth0.php [hunk 4]
```diff
@@ -254,12 +270,13 @@ public function __construct(array $config)
...
-        $this->responseType = $config['response_type'] ?? 'code';
-        $this->scope = $config['scope'] ?? 'openid profile email';
+        $this->audience      = $config['audience'] ?? null;
+        $this->responseMode  = $config['response_mode'] ?? 'query';
+        $this->responseType  = $config['response_type'] ?? 'code';
+        $this->scope         = $config['scope'] ?? 'openid profile email';
         $this->guzzleOptions = $config['guzzle_options'] ?? [];
-        $this->skipUserinfo = $config['skip_userinfo'] ?? true;
+        $this->skipUserinfo  = $config['skip_userinfo'] ?? true;
+        $this->maxAge        = $config['max_age'] ?? null;
```
Comment (joshcanhelp) [inline src/Auth0.php]: Only new line ... look for the config key or `null` if none

File: src/Auth0.php [hunk 5]
```diff
@@ -286,20 +303,22 @@ public function __construct(array $config)
...
-            if ($config['store'] === false) {
-                $emptyStore = new EmptyStore();
-                $this->setStore($emptyStore);
-            } else {
-                $this->setStore($config['store']);
-            }
-        } else {
-            $sessionStore = new SessionStore($session_base_name);
-            $this->setStore($sessionStore);
+        $userStore = $config['store'] ?? new SessionStore($session_base_name);
```
Comment (joshcanhelp) [inline src/Auth0.php]: If not set, use a `new SessionStore` as before

File: src/Auth0.php [hunk 6]
```diff
@@ -381,6 +400,8 @@ public function getLoginUrl(array $params = [])
             'response_mode' => $this->responseMode,
             'response_type' => $this->responseType,
             'redirect_uri' => $this->redirectUri,
+            'max_age' => $this->maxAge,
```
Comment (joshcanhelp) [inline src/Auth0.php]: Will be `null` if not set at the class level

File: src/Auth0.php [hunk 7]
```diff
@@ -392,6 +423,13 @@ public function getLoginUrl(array $params = [])
             $this->stateHandler->store($auth_params['state']);
         }
 
+        $auth_params['nonce'] = self::getNonce();
```
Comment (joshcanhelp) [inline src/Auth0.php]: Do not allow developers to pass in a nonce, always set one.

File: src/Auth0.php [hunk 8]
```diff
@@ -613,8 +651,15 @@ public function setIdToken($idToken)
             $sigVerifier = new SymmetricVerifier($this->clientSecret);
         }
 
+        $verifierOptions = [
+            'nonce' => $this->authStore->get('nonce'),
```
Comment (joshcanhelp) [inline src/Auth0.php]: If the stored values are empty, `null` is returned and the check is skipped

File: src/Store/CookieStore.php [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+     *
+     * @return boolean
+     */
+    protected function setCookie(string $key_name, ?string $value = null) : bool
+    {
+        if (is_null($value)) {
+            return setcookie($key_name, '', 0, '/');
+        }
+
+        return setcookie($key_name, $value, time() + $this->cookie_expiration, '/', '', false, true);
```
Comment (stevehobbsdev) [inline src/Store/CookieStore.php]: Would it make sense here to check whether the app is using TLS and set `secure` on the cookie appropriately? Or is the nonce value not really high-value enough?

File: src/Auth0.php (hunk 7 above)
Comment (lbalmaceda) [reply src/Auth0.php]: Developers could specify their own nonce if they want so. Just like they can specify a custom state as well (line above this one). If the nonce value is not present by then, that's when you generate a random one as default.

File: src/Auth0.php (hunk 8 above)
Comment (lbalmaceda) [reply src/Auth0.php]: I guess this is related to @stevehobbsdev 's question. Devs should not be able to skip this check. As far as I know, the only optional value here is max_age. For leeway you always have a default value of 60 secs, and for nonce you will either get the value set by the dev or one generated by you prior to initialize the auth. 

Please make it required. Happy to discuss internally ⚡️

File: src/Auth0.php [hunk 10]
```diff
@@ -732,4 +787,20 @@ public function setStore(StoreInterface $store)
         $this->store = $store;
         return $this;
     }
+
+    /**
+     * @param integer $length
+     *
+     * @return string
+     */
+    public static function getNonce(int $length = 16) : string
```
Comment (lbalmaceda) [inline src/Auth0.php]: Out of curiosity, this should be pretty similar to the state generation. Are you duplicating that logic or can you re-use that one instead?

File: src/Store/CookieStore.php [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+     *
+     * @param string $key   Cookie to set.
+     * @param mixed  $value Value to use.
+     *
+     * @return void
+     */
+    public function set($key, $value)
+    {
+        $key_name           = $this->getCookieName($key);
+        $_COOKIE[$key_name] = $value;
```
Comment (lbalmaceda) [inline src/Store/CookieStore.php]: what is the difference between this line and the one below?

File: src/Store/EmptyStore.php [hunk 12]
```diff
@@ -2,21 +2,8 @@
...
- */
-
-
-
 /**
  * This class is a mockup store, that discards the values, its a way of saying no store.
- *
- * @author Auth0
  */
 class EmptyStore implements StoreInterface
```
Comment (lbalmaceda) [inline src/Store/EmptyStore.php]: Assuming state should already be an enforced check, what is the reason to keep this store implementation around?

File: tests/Auth0Test.php [hunk 13]
```diff
@@ -377,7 +390,10 @@ public function testThatGetLoginUrlGeneratesState()
         $parsed_url_query = parse_url( $auth_url, PHP_URL_QUERY );
         $url_query        = explode( '&', $parsed_url_query );
 
+        $this->assertArrayHasKey( 'auth0__webauth_state', $_SESSION );
         $this->assertContains( 'state='.$_SESSION['auth0__webauth_state'], $url_query );
+        $this->assertArrayHasKey( 'auth0__nonce', $_SESSION );
```
Comment (lbalmaceda) [inline tests/Auth0Test.php]: if this is used for webauth as well, should we keep the consistency of calling it `auth0__webauth_nonce`?

File: src/Auth0.php (hunk 7 above)
Comment (joshcanhelp) [reply src/Auth0.php]: Why would a developer need to specify a nonce? State makes sense, they want that value on the return, but why nonce?

File: src/Auth0.php (hunk 8 above)
Comment (joshcanhelp) [reply src/Auth0.php]: My reply on Steve's comment below covers some of this RE: empty store option. 

The question is ... the nonce is an option in the ID token validator, not a required field. You're saying we should fail here if the nonce, say, was not stored properly and could not be retrieved? So if this is empty, fail?

File: src/Auth0.php [hunk 14]
```diff
@@ -300,6 +332,13 @@ public function __construct(array $config)
             $this->setStore($sessionStore);
         }
 
+        $authStore = $config['auth_store'] ?? new CookieStore();
+        if (! $authStore instanceof StoreInterface) {
+            $authStore = new EmptyStore();
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply src/Auth0.php]: Yes. This is a bit of a relic from the other storage option (for user identity and tokens), the storage would not error but would not store the data. 

The problem here, I'm realizing now, is that if this empty store is used, then nonce will be empty on the return and will not be checked. I'll add an actual fallback here 👍

File: src/Store/CookieStore.php (hunk 9 above)
Comment (joshcanhelp) [reply src/Store/CookieStore.php]: Not necessary, IMO. Nonce shows up in the URL so not much of a secret.

File: src/Auth0.php (hunk 10 above)
Comment (joshcanhelp) [reply src/Auth0.php]: I'll re-use, yes. I need to change how state is handled to match more closely with nonce in another PR

File: src/Store/CookieStore.php (hunk 11 above)
Comment (joshcanhelp) [reply src/Store/CookieStore.php]: `$_COOKIE` is the in-memory global of all cookie values received form the browser and `setcookie()` sends the `Set-Cookie` HTTP header

File: src/Store/EmptyStore.php (hunk 12 above)
Comment (joshcanhelp) [reply src/Store/EmptyStore.php]: Not sure but I'll evaluate in another PR. State can currently be skipped in this SDK

**Comment by lbalmaceda** [comment]
@joshcanhelp this is the main blocker on my review https://github.com/auth0/auth0-PHP/pull/395#discussion_r345757890, the rest are observations or questions. Once you fix that, it should be ready to merge provided @stevehobbsdev agrees.

I basically added that blocker because your comment stated that the dev could be able to skip checks when desired. They shouldn't be able to skip any checks 👍

File: tests/Auth0Test.php (hunk 13 above)
Comment (joshcanhelp) [reply tests/Auth0Test.php]: Discrepancies with state will be addressed in another PR. Not a big fan of the term "webauth" in this context, might get confused with [WebAuth browser API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Authentication_API)

**Comment by joshcanhelp** [comment]
@stevehobbsdev - Made a few fixes and corrections and added tests for CookieStore. I addressed all of @lbalmaceda's issues here as well.

**Pull request merged**